src/rpc/server.cpp

```diff
@@ -59,16 +59,14 @@ void RPCTypeCheck(const UniValue& params,
             break;
 
         const UniValue& v = params[i];
-        if (!(fAllowNull && v.isNull())) {
-            RPCTypeCheckArgument(v, t);
-        }
+        RPCTypeCheckArgument(v, t, fAllowNull);
         i++;
     }
 }
 
-void RPCTypeCheckArgument(const UniValue& value, const UniValueType& typeExpected)
+void RPCTypeCheckArgument(const UniValue& value, const UniValueType& typeExpected, bool allow_null)
 {
-    if (!typeExpected.typeAny && value.type() != typeExpected.type) {
+    if (!(allow_null && value.isNull()) && !typeExpected.typeAny && value.type() != typeExpected.type) {
         throw JSONRPCError(RPC_TYPE_ERROR, strprintf("Expected type %s, got %s", uvTypeName(typeExpected.type), uvTypeName(value.type())));
```

Comment:
A negative aspect of this change is that it loses the "for %s" portion of the `RPCTypeCheckObj` error message.